base/iterator.jl

```diff
@@ -298,10 +298,32 @@ done(it::Repeated, state) = false
 
 repeated(x, n::Int) = take(repeated(x), n)
 
-# product
+
+# Product -- cartesian product of iterators
 
 abstract AbstractProdIterator
+length(p::AbstractProdIterator) = prod(size(p))
+
+# one iterator
+immutable Prod1{I} <: AbstractProdIterator
+    a::I
+end
+product(a) = Prod1(a)
+
+eltype{I}(::Type{Prod1{I}}) = Tuple{eltype(I)}
+size(p::Prod1) = size(p.a)
+
+@inline start(p::Prod1) = (start(p.a),)
```

Comment:
As I said on the other issue, the function must always return a tuple to allow generic programming. See the (real) example I provided there.